File: packages/indexer/cognis_indexer/writer.py
"""IndexWriter — single-writer transaction orchestration for the indexer pipeline.

Implements task 11 of ``.kiro/specs/cognis/tasks.md``:

- 11.1  Dedicated writer logic, asyncio-friendly with one DB connection in WAL mode.
- 11.2  Per-file transaction: upsert all ``symbol``, ``edge``,
        ``symbol_attribute``, ``symbol_vec`` rows for that file atomically.
- 11.3  Deletion cascade: when a symbol disappears, ``delete_symbol``
        removes outbound edges and marks inbound edges with
        ``meta.dst_missing=true``.
- 11.4  Update ``file.parsed_at`` and ``file.parse_status`` per file.

Design references:

- *Indexer Pipeline → Writer* (design.md): "single dedicated thread …
  per-file transaction … cascade-update inverse edges on deletion".
- *Correctness Properties → CP-3*: deletion cascade invariants.

Conversion contract
-------------------
``ParsedSymbol`` (from the parser stage) is converted to a :class:`SymbolNode`
by mapping fields 1:1.  The enricher's ``untrusted_flags`` arrive already
attached to the ``ParsedSymbol`` if an enricher ran; this writer does not
re-run enrichment.

``ResolvedEdge`` (from the resolver stage) is converted to an :class:`Edge`
by forwarding ``src_id``, ``dst_id``, ``kind``, ``confidence`` and merging
``meta`` from the resolved edge (marking ``meta["ambiguous"]=true`` when
``ambiguous`` is set).

Embedding upsert
----------------
If the caller provides a non-empty ``embeddings`` dict
``{content_hash: np.ndarray}``, the writer upserts the vector for each symbol
whose ``content_hash`` has an entry.

For the ``vec0`` backend (sqlite-vec loaded), we use the native ``INSERT OR
REPLACE`` on the ``symbol_vec`` virtual table.  For the fallback table the
vector is serialised as ``numpy.ndarray.tobytes()``.

Thread safety
-------------
The :class:`Database` writer mutex (``_WRITER_LOCK``) serialises concurrent
callers through :meth:`Database.write`.  ``IndexWriter`` itself is stateless
between calls and is safe to share across coroutines/threads.
"""

from __future__ import annotations

import json
import logging
from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Any

from cognis.db import (
    Database,
    delete_symbol,
    now_epoch,
)
from cognis.models import Edge, FileRecord, SymbolAttribute, SymbolNode

from cognis_indexer.parsers.base import ParsedSymbol
from cognis_indexer.resolver.base import ResolvedEdge

# ``numpy`` ships under the ``embed-local`` extra. Import it lazily so the
# writer (and the indexer pipeline importing it) stays importable without the
# extra. ``np`` is only touched when upserting embedding vectors, which only
# happens when embeddings were produced — i.e. numpy is present.
if TYPE_CHECKING:
    import numpy as np
else:
    try:
        import numpy as np
    except ImportError:  # pragma: no cover - exercised only without embed-local
        np = None

__all__ = ["FileWritePayload", "IndexWriter"]

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Public payload dataclass
# ---------------------------------------------------------------------------


@dataclass
class FileWritePayload:
    """Complete write payload for one file pass through the indexer pipeline.

    All fields are required by the writer for a complete atomic file update.
    ``embeddings`` may be empty when the embedder was skipped (``--skip-embeddings``).
    """

    file_path: str
    """Repo-relative path, forward slashes (matches ``file.path`` PK)."""

    language: str
    """Language identifier, e.g. ``"python"``, ``"typescript"``, ``"go"``."""

    file_size_bytes: int
    """Raw byte count of the source file (not content length)."""

    content_hash: str
    """``sha256(file_content_bytes)[:16]`` — used for file-level dedup."""

    parsed_at: int
    """Unix epoch when parsing finished; use :func:`cognis.db.now_epoch`."""

    parse_status: str  # "ok" | "partial" | "failed"
    """Outcome of the parse pass (mirrors ``FileRecord.parse_status``)."""

    symbols: list[ParsedSymbol] = field(default_factory=list)
    """All symbols extracted from this file in this pass."""

    edges: list[ResolvedEdge] = field(default_factory=list)
    """Call/import/etc. edges resolved for symbols in this file."""

    attributes: list[SymbolAttribute] = field(default_factory=list)
    """Enricher-extracted side-effect attributes (db_table, http_route, …)."""

    embeddings: dict[str, np.ndarray] = field(default_factory=dict)
    """``{content_hash: vector}`` — optional.  Empty → no vec upsert."""


# ---------------------------------------------------------------------------
# Conversion helpers
# ---------------------------------------------------------------------------


def _parsed_to_node(sym: ParsedSymbol) -> SymbolNode:
    """Convert a :class:`ParsedSymbol` to a :class:`SymbolNode` for DB insertion."""
    return SymbolNode(
        id=sym.id,
        kind=sym.kind,
        name=sym.name,
        qualified_name=sym.qualified_name,
        language=sym.language,
        module=sym.module,
        file_path=sym.file_path,
        line_start=sym.line_start,
        line_end=sym.line_end,
        signature=sym.signature,
        docstring=sym.docstring,
        content_hash=sym.content_hash,
        body_excerpt=sym.body_excerpt,
        semantic_summary=None,  # filled later by summariser (Phase 2)
        risk_score=0.0,
        ambiguous=False,
        untrusted_flags=list(sym.untrusted_flags),
        updated_at=now_epoch(),
    )


def _symbol_ids_for_file(db: Database, file_path: str) -> set[str]:
    """Return the current symbol ids for *file_path* without scanning the full DB."""
    conn = db.connect()
    rows = conn.execute(
        "SELECT id FROM symbol WHERE file_path = ?",
        (file_path,),
    ).fetchall()
    return {str(row["id"]) for row in rows}


def _resolved_to_edge(re: ResolvedEdge) -> Edge:
    """Convert a :class:`ResolvedEdge` to a :class:`Edge` for DB insertion.

    Merges the ``ambiguous`` flag into ``meta`` so it is preserved in the DB.
    """
    meta: dict[str, Any] = dict(re.meta)
    if re.ambiguous:
        meta["ambiguous"] = True
    return Edge(
        src_id=re.src_id,
        dst_id=re.dst_id,
        kind=re.kind,
        confidence=re.confidence,
        meta=meta,
    )


# ---------------------------------------------------------------------------
# Embedding upsert helper
# ---------------------------------------------------------------------------


def _upsert_embeddings(
    db: Database,
    symbols: list[SymbolNode],
    embeddings: dict[str, np.ndarray],
) -> None:
    """Upsert symbol vectors — handles both vec0 and fallback table.

    Only symbols whose ``content_hash`` appears in *embeddings* are updated.
    This keeps the function a pure side-effect over the mapping; callers
    decide what to embed.
    """
    if not embeddings:
        return

    # Last write wins when the parser emitted duplicate ids for one file batch.
    by_id: dict[str, Any] = {}
    for sym in symbols:
        vec = embeddings.get(sym.content_hash)
        if vec is None:
            continue
        by_id[sym.id] = vec

    if not by_id:
        return

    payload = [(sym_id, vec.astype(np.float32).tobytes()) for sym_id, vec in by_id.items()]
    with db.write() as conn:
        if db.vec_enabled:
            # vec0 does not support INSERT OR REPLACE (sqlite-vec #259).
            # Delete+insert per row avoids UNIQUE failures from soft-deleted rows
            # and duplicate ids in a single executemany batch.
            for sym_id, vec_bytes in payload:
                conn.execute("DELETE FROM symbol_vec WHERE symbol_id = ?", (sym_id,))
                conn.execute(
                    "INSERT INTO symbol_vec(symbol_id, embedding) VALUES (?, ?)",
                    (sym_id, vec_bytes),
                )
        else:
            conn.executemany(
                "INSERT OR REPLACE INTO symbol_vec(symbol_id, embedding) VALUES (?, ?)",
                payload,
            )


# ---------------------------------------------------------------------------
# IndexWriter
# ---------------------------------------------------------------------------


class IndexWriter:
    """Orchestrates atomic per-file writes to the UCKG database.

    Usage::

        writer = IndexWriter(db)
        await writer.write_file(payload)
        await writer.delete_file("src/old.py")
        writer.close()

    The async interface is provided for compatibility with the broader async
    indexer pipeline.  The underlying DB I/O is synchronous (SQLite WAL) but
    the ``await``-able signature allows callers to ``await`` without blocking the
    event loop in environments where the writer runs in a thread executor.

    In practice the writer executes synchronously within the asyncio event loop
    — this is acceptable for the MVP because SQLite WAL writes are fast and the
    design allocates a dedicated indexer process (``cognis-indexd``) so there is
    no MCP query path blocked by writer I/O.
    """

    def __init__(self, db: Database) -> None:
        self._db = db

    # ------------------------------------------------------------------
    # Public async API
    # ------------------------------------------------------------------

    async def write_file(self, payload: FileWritePayload) -> None:
        """Write an entire file payload atomically (one transaction).

        Steps (design 11.2):

        1. Query the DB for symbols currently associated with *file_path*
           (the "old" set from the previous parse).
        2. Upsert all new/changed symbols.
        3. Delete symbols that were in the old set but not the new set
           (cascade per CP-3 / 11.3).
        4. Upsert edges from the resolver output.
        5. Upsert ``symbol_attribute`` rows from the enricher.
        6. Upsert ``symbol_vec`` rows if embeddings provided (11.2f).
        7. Upsert the ``file`` row with updated ``parsed_at`` / ``parse_status``
           (11.4).
        """
        self._write_file_sync(payload)

    async def delete_file(self, file_path: str) -> None:
        """Remove all symbols for *file_path* and apply cascade (11.3).

        Also removes the ``file`` row.  Idempotent — deleting a file that
        never existed is a no-op.
        """
        self._delete_file_sync(file_path)

    def close(self) -> None:
        """Release the writer's DB connection on the calling thread."""
        self._db.close_thread_connection()

    # ------------------------------------------------------------------
    # Sync implementations (separated for testability)
    # ------------------------------------------------------------------

    def _write_file_sync(self, payload: FileWritePayload) -> set[str]:
        """Synchronous core of :meth:`write_file`.

        Returns:
            Set of symbol ids removed from the file during this rewrite.
        """
        db = self._db

        # Step 1: discover old symbol ids for this file without materializing
        # the whole symbol table. Full-DB scans here become prohibitive on large
        # repos because this path runs once per file write.
        old_ids = _symbol_ids_for_file(db, payload.file_path)

        # Build new symbol nodes from ParsedSymbol list.
        new_nodes = [_parsed_to_node(s) for s in payload.symbols]
        new_ids = {n.id for n in new_nodes}

        # Step 2 + 3 + 4 + 5 + 7 — all in one atomic write block.
        with db.write() as conn:
            # 2. Upsert new/changed symbols.
            _upsert_symbols_conn(conn, new_nodes)

            # 3. Delete removed symbols (CP-3 cascade via delete_symbol helper).
            # We call delete_symbol per-symbol because it applies the edge
            # cascade.  To stay within the current open transaction we use
            # the raw SQL operations directly (delete_symbol opens its own
            # transaction).  Instead we inline the cascade SQL here.
            removed_ids = old_ids - new_ids
            for sym_id in removed_ids:
                conn.execute("DELETE FROM symbol WHERE id = ?", (sym_id,))
                conn.execute("DELETE FROM edge WHERE src_id = ?", (sym_id,))
                conn.execute(
                    """
                    UPDATE edge
                       SET meta = json_patch(
                               COALESCE(meta, '{}'),
                               json_object('dst_missing', json('true'))
                           )
                     WHERE dst_id = ?
                    """,
                    (sym_id,),
                )

            # 4. Upsert edges.
            edge_rows = [_resolved_to_edge(re) for re in payload.edges]
            _upsert_edges_conn(conn, edge_rows)

            # 5. Upsert symbol attributes.
            _upsert_attributes_conn(conn, payload.attributes)

            # 7. Upsert file row.
            file_record = FileRecord(
                path=payload.file_path,
                language=payload.language,
                size_bytes=payload.file_size_bytes,
                content_hash=payload.content_hash,
                parsed_at=payload.parsed_at,
                parse_status=payload.parse_status,  # type: ignore[arg-type]
            )
            _upsert_file_conn(conn, file_record)

        # Step 6: embeddings upsert (separate transaction — vector table may
        # be a separate vtable and sqlite-vec doesn't support multi-statement
        # transactions in all versions).
        _upsert_embeddings(db, new_nodes, payload.embeddings)
        return removed_ids

    def _delete_file_sync(self, file_path: str) -> None:
        """Synchronous core of :meth:`delete_file`."""
        db = self._db

        # Delete each symbol applying cascade (outbound edges deleted,
        # inbound edges flagged dst_missing).
        for sym_id in _symbol_ids_for_file(db, file_path):
            delete_symbol(db, sym_id)

        # Remove the file row.
        with db.write() as conn:
            conn.execute("DELETE FROM file WHERE path = ?", (file_path,))


# ---------------------------------------------------------------------------
# Connection-level helpers (inline within an open transaction)
# ---------------------------------------------------------------------------

_SYMBOL_COLUMNS = (
    "id",
    "kind",
    "name",
    "qualified_name",
    "language",
    "module",
    "file_path",
    "line_start",
    "line_end",
    "signature",
    "docstring",
    "content_hash",
    "body_excerpt",
    "semantic_summary",
    "risk_score",
    "ambiguous",
    "untrusted_flags",
    "updated_at",
)


def _symbol_to_row(sym: SymbolNode) -> tuple[Any, ...]:
    return (
        sym.id,
        sym.kind,
        sym.name,
        sym.qualified_name,
        sym.language,
        sym.module,
        sym.file_path,
        sym.line_start,
        sym.line_end,
        sym.signature,
        sym.docstring,
        sym.content_hash,
        sym.body_excerpt,
        sym.semantic_summary,
        sym.risk_score,
        1 if sym.ambiguous else 0,
        json.dumps(sym.untrusted_flags) if sym.untrusted_flags else None,
        sym.updated_at,
    )


def _upsert_symbols_conn(conn: Any, symbols: list[SymbolNode]) -> None:
    """Upsert symbols on an already-open connection (no new transaction)."""
    if not symbols:
        return
    placeholders = ", ".join(["?"] * len(_SYMBOL_COLUMNS))
    update_clause = ", ".join(f"{col} = excluded.{col}" for col in _SYMBOL_COLUMNS if col != "id")
    sql = (
        f"INSERT INTO symbol ({', '.join(_SYMBOL_COLUMNS)}) "
        f"VALUES ({placeholders}) "
        f"ON CONFLICT(id) DO UPDATE SET {update_clause}"
    )
    conn.executemany(sql, [_symbol_to_row(s) for s in symbols])


def _upsert_edges_conn(conn: Any, edges: list[Edge]) -> None:
    """Upsert edges on an already-open connection."""
    if not edges:
        return
    rows = [
        (e.src_id, e.dst_id, e.kind, e.confidence, json.dumps(e.meta) if e.meta else None)
        for e in edges
    ]
    conn.executemany(
        "INSERT INTO edge (src_id, dst_id, kind, confidence, meta) "
        "VALUES (?, ?, ?, ?, ?) "
        "ON CONFLICT(src_id, dst_id, kind) DO UPDATE SET "
        "confidence = excluded.confidence, meta = excluded.meta",
        rows,
    )


def _upsert_attributes_conn(conn: Any, attrs: list[SymbolAttribute]) -> None:
    """Upsert symbol attributes on an already-open connection."""
    if not attrs:
        return
    conn.executemany(
        "INSERT OR REPLACE INTO symbol_attribute (symbol_id, key, value) VALUES (?, ?, ?)",
        [(a.symbol_id, a.key, a.value) for a in attrs],
    )


def _upsert_file_conn(conn: Any, record: FileRecord) -> None:
    """Upsert a file row on an already-open connection."""
    conn.execute(
        "INSERT INTO file (path, language, size_bytes, content_hash, parsed_at, parse_status) "
        "VALUES (?, ?, ?, ?, ?, ?) "
        "ON CONFLICT(path) DO UPDATE SET "
        "language = excluded.language, "
        "size_bytes = excluded.size_bytes, "
        "content_hash = excluded.content_hash, "
        "parsed_at = excluded.parsed_at, "
        "parse_status = excluded.parse_status",
        (
            record.path,
            record.language,
            record.size_bytes,
            record.content_hash,
            record.parsed_at,
            record.parse_status,
        ),
    )

File: crates/cognis-indexer/src/writer.rs
//! Writer stage of the indexer pipeline (Task 8.2).
//!
//! Persists one file's symbols and resolved edges through the
//! [`SymbolWriter`](cognis_store::SymbolWriter) trait. Rust mirror of
//! `cognis_indexer.writer.IndexWriter`'s per-file write path, scoped to the
//! symbol + edge surface the trait exposes (the `file` cache row + incremental
//! removal diff land with the watcher in Task 8.3; embeddings are the
//! embedder's surface).
//!
//! **Per-file transaction:** each [`SymbolWriter`] call runs under its own
//! `BEGIN IMMEDIATE` transaction in `cognis-store`, so a file's symbols commit
//! atomically and its edges commit atomically. Symbols are written **before**
//! edges so referenced rows exist first (the `edge` FK is application-layer, so
//! this is ordering hygiene rather than a hard constraint).
//!
//! **Secret scrub before persist (Requirement 9.3):** this stage persists
//! whatever the [`Enricher`](crate::enricher::Enricher) handed it. Callers MUST
//! enrich symbols before constructing a [`FileWritePayload`]; the redaction
//! therefore always happens upstream of any DB write.
//!
//! **`meta.dst_missing` convention:** when a symbol is removed,
//! [`IndexWriter::delete_symbols`] delegates to `SymbolWriter::delete_symbol`,
//! which deletes outbound edges and flags inbound edges `meta.dst_missing =
//! true` rather than erasing them — preserving the audit trail.

use std::time::{SystemTime, UNIX_EPOCH};

use cognis_core::{Edge, Result, Symbol};
use cognis_store::SymbolWriter;

/// Everything the Writer needs to persist one file's parse pass.
///
/// Symbols are expected to be **already enriched** (secrets scrubbed,
/// `untrusted_flags` set — Requirement 9.3) and edges already converted from
/// the resolver output (see [`crate::resolver::to_edges`]).
#[derive(Debug, Clone, Default)]
pub struct FileWritePayload {
    /// Enriched symbols extracted from this file in this pass.
    pub symbols: Vec<Symbol>,
    /// Resolved edges whose `src_id` belongs to this file.
    pub edges: Vec<Edge>,
}

impl FileWritePayload {
    /// Construct a payload from enriched symbols and resolved edges.
    pub fn new(symbols: Vec<Symbol>, edges: Vec<Edge>) -> Self {
        FileWritePayload { symbols, edges }
    }
}

/// Persists symbol/edge batches through any [`SymbolWriter`] backend.
///
/// Generic over the writer so it works against `cognis_store::Database` in
/// production and any in-memory `SymbolWriter` in tests.
#[derive(Debug)]
pub struct IndexWriter<W: SymbolWriter> {
    writer: W,
}

impl<W: SymbolWriter> IndexWriter<W> {
    /// Wrap a [`SymbolWriter`] backend.
    pub fn new(writer: W) -> Self {
        IndexWriter { writer }
    }

    /// Consume the writer, returning the underlying backend.
    pub fn into_inner(self) -> W {
        self.writer
    }

    /// Persist one file's payload: upsert symbols, then upsert edges.
    ///
    /// Each symbol's `updated_at` is stamped with the current epoch second at
    /// write time (mirror the Python writer's `now_epoch()` on every write), so
    /// re-indexing an unchanged file still advances the freshness marker.
    /// Returns early without touching the DB when the payload is empty.
    pub fn write_file(&mut self, payload: &FileWritePayload) -> Result<()> {
        if payload.symbols.is_empty() && payload.edges.is_empty() {
            return Ok(());
        }

        if !payload.symbols.is_empty() {
            let now = now_epoch();
            let mut stamped = payload.symbols.clone();
            for s in &mut stamped {
                s.updated_at = now;
            }
            self.writer.upsert_symbols(&stamped)?;
        }

        if !payload.edges.is_empty() {
            self.writer.upsert_edges(&payload.edges)?;
        }

        Ok(())
    }

    /// Remove each id in `ids`, applying the delete cascade (outbound edges
    /// deleted, inbound edges flagged `meta.dst_missing = true`). Used by the
    /// incremental path when a file's symbol set shrinks between passes.
    pub fn delete_symbols(&mut self, ids: &[String]) -> Result<()> {
        for id in ids {
            self.writer.delete_symbol(id)?;
        }
        Ok(())
    }
}

/// Current Unix epoch in whole seconds (mirror `cognis.db.now_epoch`).
fn now_epoch() -> i64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs() as i64)
        .unwrap_or(0)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::enricher::Enricher;
    use crate::parser::parse_source;
    use crate::resolver::{resolve_edges, to_edges};
    use cognis_store::Database;

    fn mem_db() -> Database {
        // Per-thread in-memory DB; a test runs on one thread so the cached
        // connection is reused across reader/writer handles.
        Database::open(":memory:").expect("open in-memory db")
    }

    #[test]
    fn writes_symbols_and_edges_per_file() {
        let out = parse_source(
            "m.py",
            "def helper():\n    return 1\n\ndef caller():\n    return helper()\n",
        );
        let enricher = Enricher::new();
        let symbols: Vec<Symbol> = out
            .symbols
            .iter()
            .map(|s| enricher.enrich(s).symbol)
            .collect();
        let edges = to_edges(&resolve_edges(&symbols));
        assert!(!edges.is_empty(), "expected a caller->helper edge");

        let db = mem_db();
        let mut writer = IndexWriter::new(db.clone());
        writer
            .write_file(&FileWritePayload::new(symbols.clone(), edges.clone()))
            .expect("write_file");

        // Symbols + edges are now persisted and readable.
        assert_eq!(db.list_symbols().unwrap().len(), symbols.len());
        assert_eq!(db.list_edges().unwrap().len(), edges.len());
        // updated_at was stamped (non-zero) at write time.
        assert!(db.list_symbols().unwrap().iter().all(|s| s.updated_at > 0));
    }

    #[test]
    fn redacted_body_is_what_lands_in_db() {
        let src = "def connect():\n    password = \"hunter2secret\"\n    return password\n";
        let out = parse_source("m.py", src);
        let enricher = Enricher::new();
        let symbols: Vec<Symbol> = out
            .symbols
            .iter()
            .map(|s| enricher.enrich(s).symbol)
            .collect();

        let db = mem_db();
        let mut writer = IndexWriter::new(db.clone());
        writer
            .write_file(&FileWritePayload::new(symbols, Vec::new()))
            .expect("write_file");

        // Requirement 9.3: the secret never reaches the DB.
        for s in db.list_symbols().unwrap() {
            let body = s.body_excerpt.unwrap_or_default();
            assert!(!body.contains("hunter2secret"));
        }
    }

    #[test]
    fn delete_symbol_flags_inbound_edge_dst_missing() {
        let out = parse_source(
            "m.py",
            "def helper():\n    return 1\n\ndef caller():\n    return helper()\n",
        );
        let symbols: Vec<Symbol> = out.symbols.clone();
        let edges = to_edges(&resolve_edges(&symbols));
        let helper_id = symbols
            .iter()
            .find(|s| s.name == "helper")
            .unwrap()
            .id
            .clone();

        let db = mem_db();
        let mut writer = IndexWriter::new(db.clone());
        writer
            .write_file(&FileWritePayload::new(symbols, edges))
            .expect("write");

        writer
            .delete_symbols(std::slice::from_ref(&helper_id))
            .expect("delete");

        // Inbound edge to helper is kept but flagged dst_missing.
        let inbound: Vec<Edge> = db
            .list_edges()
            .unwrap()
            .into_iter()
            .filter(|e| e.dst_id == helper_id)
            .collect();
        assert!(!inbound.is_empty(), "inbound edge should be preserved");
        assert!(inbound.iter().all(|e| e.dst_missing()));
    }

    #[test]
    fn empty_payload_is_noop() {
        let db = mem_db();
        let mut writer = IndexWriter::new(db.clone());
        writer
            .write_file(&FileWritePayload::default())
            .expect("noop");
        assert_eq!(db.list_symbols().unwrap().len(), 0);
    }
}
